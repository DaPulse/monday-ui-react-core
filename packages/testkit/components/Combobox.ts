import { Locator, Page } from "@playwright/test";
import { BaseElement } from "./BaseElement";
import { Search } from "./Search";
import { ListItem } from "./ListItem";

export class Combobox extends BaseElement {
  private searchInput: Search;
  /**
   * Create a Combobox.
   * @param {Page} page - The Playwright page object.
   * @param {Locator} locator - The locator for the Combobox element.
   * @param {string} elementReportName - The name for reporting purposes.
   */
  constructor(page: Page, locator: Locator, elementReportName: string) {
    super(page, locator, elementReportName);
    this.searchInput = new Search(page, this.locator.locator("[role='search']"), `${elementReportName} Search Input`);
  }

  /**
   * Select an item from the combobox.
   * @param {string} item - The name of the item to select.
   * @returns {Promise<void>}
   */
  async selectItem(item: string): Promise<void> {
    await this.searchInput.setText(item);
    const comboBoxItem = new ListItem(
      this.page,
      this.locator.locator(`[role='listbox']:has-text("${item}")`),
      `Select Combobox Item: ${item}`
    );
    await comboBoxItem.click();
  }
}
