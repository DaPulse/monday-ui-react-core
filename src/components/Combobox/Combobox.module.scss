@import "../../styles/themes";
@import "../../styles/typography";
@import "../../styles/states";

.comboboxWrapper {
  height: 100%;
  display: flex;
  flex-direction: column;
  @include font-general-text();
  position: relative;
  margin-bottom: var(--spacing-small);
}

.comboboxWrapper .comboboxCategoryWithDivider {
  margin-top: var(--spacing-medium);
}

.comboboxWrapper.stickyCategory .comboboxCategory {
  z-index: 10;
  background-color: var(--secondary-background-color);
}

/* Relative to Search height */

.comboboxWrapper.stickyCategory.sizeSmall .comboboxCategory {
  top: 32px;
}

.comboboxWrapper.stickyCategory.sizeMedium .comboboxCategory {
  top: 40px;
}

.comboboxWrapper.stickyCategory.sizeLarge .comboboxCategory {
  top: 48px;
}

.comboboxWrapper.empty {
  max-height: 200px;
}

.comboboxWrapperSearchWrapper {
  position: sticky;
  top: 0;
  z-index: 12;
}

.comboboxWrapperSearch {
  flex: 0 0 auto;
}

.comboboxWrapperNoResults {
  margin-top: var(--spacing-small);
  flex: 1 1 auto;
  height: calc(100% - 20px);
  overflow: hidden;
  position: relative;
  padding: 0 16px;
  display: flex;
  flex-direction: column;
}

.comboboxWrapperNoResults .comboboxMessageWrapper {
  flex: 1 0 auto;
  display: flex;
  align-items: center;
  justify-content: center;
}

.comboboxWrapperNoResults .comboboxMessageWrapper .comboboxMessage {
  @include ellipsis();
}

.comboboxWrapperNoResults .addNewButton {
  flex: 0 0 auto;
  width: 100% !important;
  margin-bottom: 8px;
}

.comboboxWrapperNoResults .addNewButton .buttonLabel {
  @include ellipsis();
}

.comboboxWrapperList {
  display: flex;
  flex-direction: column;
  flex: 1 1 auto;
  padding: 0 4px;
  overflow: hidden;
}
