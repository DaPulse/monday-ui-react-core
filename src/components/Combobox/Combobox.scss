@import "../../styles/themes.scss";
@import "../../styles/typography.scss";
@import "../../styles/states.scss";

$icon-margin: 4px;

.combobox--wrapper {
  overflow: hidden;
  height: 100%;
  display: flex;
  flex-direction: column;
  @include font-general-text();

  &.empty {
    max-height: 200px;
  }

  &-search {
    flex: 0 0 auto;
  }

  &-no-results {
    flex: 1 1 auto;
    height: 100%;
    overflow: hidden;
    position: relative;
    padding: 0 16px;
    display: flex;
    flex-direction: column;

    .message-wrapper {
      flex: 1 0 auto;
      display: flex;
      align-items: center;
      justify-content: center;
      .message {
        @include ellipsis();
      }
    }

    .add-new-button {
      flex: 0 0 auto;
      width: 100% !important;
      margin-bottom: 8px;

      .button-label {
        @include ellipsis();
      }
    }
  }

  &-list {
    flex: 1 1 auto;
    padding: 8px;

    .combobox-option {
      padding: 8px;
      height: 32px;
      border-radius: 4px;
      display: flex;
      align-items: center;
      overflow: hidden;
      cursor: pointer;
      user-select: none;

      &.disabled {
        cursor: not-allowed;
        @include disabled;
      }

      &.selected {
        @include theme-prop(background-color, primary-selected-color);
      }

      &:hover:not(.disabled):not(.selected),
      &.active {
        @include theme-prop(color, primary-text-color);
        @include theme-prop(background-color, primary-background-hover-color);
      }

      &.active {
        position: relative;
        &.active-outline {
          @include focus-style-css();
        }
      }

      .option-label {
        flex: 1 1 auto;
      }

      .option-icon {
        flex: 0 0 auto;
        &.left {
          margin-right: $icon-margin;
        }
      }
    }
  }
}
