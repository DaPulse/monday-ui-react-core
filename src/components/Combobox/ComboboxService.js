export const getOptionsByCategories = (options, categories, filterValue) => {
  return options.reduce((result, option) => {
    const categoryId = option.categoryId;
    // skipping if the option doesn't have a category
    if (!categoryId) return result;
    if (categories[categoryId]?.onlyShowOnSearch && !filterValue) return result;

    if (result[categoryId]) {
      result[categoryId].push(option);
    } else {
      // eslint-disable-next-line no-param-reassign
      result[categoryId] = [option];
    }

    return result;
  }, {});
};
