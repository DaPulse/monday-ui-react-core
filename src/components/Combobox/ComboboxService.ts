import { ComboboxCategoryMap, ComboboxOptionType } from "./components/ComboboxConstants";

type OptionsByCategories = {
  [key: string]: Array<ComboboxOptionType>;
};

export const getOptionsByCategories = (
  options: ComboboxOptionType[],
  categories: ComboboxCategoryMap,
  filterValue: string
): OptionsByCategories => {
  const optionsByCategories = options.reduce((result: OptionsByCategories, option) => {
    const categoryId = option.categoryId;
    // skipping if the option doesn't have a category
    if (!categoryId) return result;
    if (categories[categoryId]?.onlyShowOnSearch && !filterValue) return result;

    if (result[categoryId]) {
      result[categoryId].push(option);
    } else {
      // eslint-disable-next-line no-param-reassign
      result[categoryId] = [option];
    }

    return result;
  }, {});

  // reorder the keys according to the categories order
  return Object.keys(categories).reduce((result: OptionsByCategories, categoryId) => {
    // eslint-disable-next-line no-param-reassign
    if (optionsByCategories[categoryId]) result[categoryId] = optionsByCategories[categoryId];

    return result;
  }, {});
};

export const defaultFilter = (filterValue: string, options: ComboboxOptionType[]) =>
  options.filter(
    ({ label }: { label: string }) => !filterValue || label.toLowerCase().includes(filterValue.toLowerCase())
  );
