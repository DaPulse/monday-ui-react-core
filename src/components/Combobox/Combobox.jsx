/* eslint-disable jsx-a11y/click-events-have-key-events */
import React, { useRef, useState, forwardRef, useMemo, useCallback } from "react";
import PropTypes from "prop-types";
import isFunction from "lodash/isFunction";
import NOOP from "lodash/noop";
import cx from "classnames";
import useMergeRefs from "../../hooks/useMergeRefs";
import Search from "../Search/Search";
import { SIZES } from "../../constants/sizes";
import Button from "../Button/Button";
import useListKeyboardNavigation from "../../hooks/useListKeyboardNavigation";
import ComboboxOption from "./components/ComboboxOption/ComboboxOption";
import ComboboxCategory from "./components/ComboboxCategory/ComboboxCategory";
import { getOptionsByCategories, defaultFilter } from "./ComboboxService";
import "./Combobox.scss";
import Divider from "../Divider/Divider";

const Combobox = forwardRef(
  (
    {
      className,
      id,
      placeholder,
      size,
      optionLineHeight,
      optionsListHeight,
      autoFocus,
      disabled,
      options,
      categories,
      withCategoriesDivider,
      noResultsMessage,
      onAddNew,
      addNewLabel,
      onClick,
      filter,
      disableFilter,
      onFilterChanged,
      loading,
      onOptionHover,
      onOptionLeave,
      shouldScrollToSelectedItem,
      noResultsRenderer,
      stickyCategories,
      optionRenderer,
      clearFilterOnSelection
    },
    ref
  ) => {
    const componentRef = useRef(null);
    const inputRef = useRef(null);
    const [activeItemIndex, setActiveItemIndex] = useState(-1);
    const [isActiveByKeyboard, setIsActiveByKeyboard] = useState(false);
    const [filterValue, setFilterValue] = useState("");
    const mergedRef = useMergeRefs({ refs: [ref, componentRef] });

    const setActiveItemIndexKeyboardNav = useCallback(
      index => {
        setActiveItemIndex(index);
        setIsActiveByKeyboard(true);
      },
      [setActiveItemIndex]
    );

    const onOptionClick = useCallback(
      (_event, index, option, mouseClick) => {
        if (option.disabled) return;
        onClick && onClick(option);
        setActiveItemIndex(index);
        if (mouseClick) {
          // set focus on input again
          requestAnimationFrame(() => inputRef.current?.focus());
        }
        setIsActiveByKeyboard(!mouseClick);
        if (clearFilterOnSelection) {
          // clear filter after adding
          onChangeCallback("");
        }
      },
      [onClick, onChangeCallback, clearFilterOnSelection]
    );

    const onOptionEnter = useCallback(
      (event, index, option) => {
        setActiveItemIndex(-1);
        onOptionHover(event, index, option);
        if (clearFilterOnSelection) {
          // clear filter after adding
          onChangeCallback("");
        }
      },
      [setActiveItemIndex, onOptionHover, onChangeCallback, clearFilterOnSelection]
    );

    const filteredOptions = useMemo(() => {
      if (disableFilter) {
        return options;
      }
      return filter(filterValue, options);
    }, [options, filterValue, filter, disableFilter]);

    const renderedItems = useMemo(() => {
      if (categories) {
        const optionsByCategories = getOptionsByCategories(filteredOptions, categories, filterValue);

        let index = 0;
        return Object.keys(optionsByCategories).map((categoryId, categoryIndex) => {
          const withDivider = withCategoriesDivider && categoryIndex !== 0;
          return [
            withDivider ? <Divider className="combobox_category-divider" key={`${categoryId}-divider`} /> : null,
            <div
              role="group"
              aria-labelledby={`combox-category-${categoryId}`}
              key={categoryId}
              className={cx({ ["combobox_category--with-divider"]: withDivider })}
            >
              <ComboboxCategory category={categories[categoryId]} />
              {optionsByCategories[categoryId].map(option => {
                const renderedOption = (
                  <ComboboxOption
                    index={index}
                    key={option.id || index}
                    option={option}
                    optionRenderer={optionRenderer}
                    isActive={activeItemIndex === index}
                    isActiveByKeyboard={isActiveByKeyboard}
                    onOptionClick={onOptionClick}
                    onOptionHover={onOptionEnter}
                    onOptionLeave={onOptionLeave}
                    optionLineHeight={optionLineHeight}
                    shouldScrollWhenActive={shouldScrollToSelectedItem}
                  />
                );
                index++;
                return renderedOption;
              })}
            </div>
          ];
        });
      }
      return filteredOptions.map((option, index) => {
        return (
          <ComboboxOption
            index={index}
            key={option.id || index}
            option={option}
            optionRenderer={optionRenderer}
            isActive={activeItemIndex === index}
            isActiveByKeyboard={isActiveByKeyboard}
            onOptionClick={onOptionClick}
            onOptionHover={onOptionEnter}
            onOptionLeave={onOptionLeave}
            optionLineHeight={optionLineHeight}
            shouldScrollWhenActive={shouldScrollToSelectedItem}
          />
        );
      });
    }, [
      categories,
      filteredOptions,
      filterValue,
      withCategoriesDivider,
      optionRenderer,
      activeItemIndex,
      isActiveByKeyboard,
      onOptionClick,
      onOptionEnter,
      onOptionLeave,
      optionLineHeight,
      shouldScrollToSelectedItem
    ]);

    const onChangeCallback = useCallback(
      value => {
        if (onFilterChanged) {
          onFilterChanged(value);
        }
        setFilterValue(value);
      },
      [setFilterValue, onFilterChanged]
    );

    const isChildSelectable = useCallback(option => {
      return option && !option.disabled;
    }, []);

    const onAddNewCallback = useCallback(() => {
      onAddNew && onAddNew(filterValue);
      // clear filter after adding
      setFilterValue("");
    }, [onAddNew, filterValue, setFilterValue]);

    useListKeyboardNavigation(
      inputRef,
      filteredOptions,
      activeItemIndex,
      setActiveItemIndexKeyboardNav,
      isChildSelectable,
      onOptionClick
    );

    const hasResults = renderedItems.length > 0;
    const hasFilter = filterValue.length > 0;

    function getAddNewLabel() {
      if (isFunction(addNewLabel)) {
        return addNewLabel(filterValue);
      }
      return addNewLabel;
    }

    function renderNoResults() {
      if (noResultsRenderer) {
        return noResultsRenderer();
      }

      return (
        <div className="combobox--wrapper-no-results">
          <div className="message-wrapper">
            <span className="message">{noResultsMessage}</span>
          </div>
          {onAddNew && !disabled && (
            <Button className="add-new-button" size={size} kind={Button.kinds.TERTIARY} onClick={onAddNewCallback}>
              <span className="button-label">{getAddNewLabel()}</span>
            </Button>
          )}
        </div>
      );
    }

    return (
      // eslint-disable-next-line jsx-a11y/aria-activedescendant-has-tabindex
      <div
        ref={mergedRef}
        className={cx("combobox--wrapper", className, `size-${size}`, {
          empty: !hasResults,
          "sticky-category": stickyCategories
        })}
        id={id}
      >
        <div className="combobox--wrapper-list" style={{ maxHeight: optionsListHeight }} role="listbox">
          <Search
            ref={inputRef}
            value={filterValue}
            wrapperClassName="combobox--wrapper-search-wrapper"
            className="combobox--wrapper-search"
            inputAriaLabel="Search for content"
            activeDescendant={`combobox-item-${activeItemIndex}`}
            id="combobox-search"
            placeholder={placeholder}
            size={size}
            disabled={disabled}
            onChange={onChangeCallback}
            autoFocus={autoFocus}
            loading={loading}
          />
          {renderedItems}
        </div>
        {hasFilter && !hasResults && !loading && renderNoResults()}
      </div>
    );
  }
);

// Locate loading next to search icon
// color it with --secondary-text-color
// size it like the icon - we think it's 16px - make sure it's not fat

Combobox.sizes = SIZES;
Combobox.iconTypes = ComboboxOption.iconTypes;

Combobox.propTypes = {
  className: PropTypes.string,
  /**
   * Placeholder to show when no value was selected
   */
  placeholder: PropTypes.string,
  /**
   * Unique element id
   */
  id: PropTypes.string,
  /**
   * A message that will be displayed inside the combo box when no results are found
   */
  noResultsMessage: PropTypes.string,
  disabled: PropTypes.bool,
  options: PropTypes.arrayOf(PropTypes.object),
  categories: PropTypes.object,
  /**
   * Divider between categories sections
   */
  withCategoriesDivider: PropTypes.bool,
  size: PropTypes.oneOf([Combobox.sizes.SMALL, Combobox.sizes.MEDIUM, Combobox.sizes.LARGE]),
  optionLineHeight: PropTypes.number,
  optionsListHeight: PropTypes.number,
  autoFocus: PropTypes.bool,
  onAddNew: PropTypes.func,
  /**
   * The label of the button that appears at the end of the combo box when the search does not return appropriate options
   */
  addNewLabel: PropTypes.oneOfType([PropTypes.string, PropTypes.func]),
  filter: PropTypes.func,
  disableFilter: PropTypes.bool,
  onFilterChanged: PropTypes.func,
  /**
   * Display the combo box with loading state
   */
  loading: PropTypes.bool,
  /**
   * on mouse hover callback for option
   */
  onOptionHover: PropTypes.func,
  /**
   * on mouse leave callback for option
   */
  onOptionLeave: PropTypes.func,
  /**
   * Allowed to the following behavior: scrolling automatically to the combo box's selected option
   */
  shouldScrollToSelectedItem: PropTypes.bool,
  noResultsRenderer: PropTypes.func,
  stickyCategories: PropTypes.bool,
  /** Clear the filter/search on selection (click or enter) */
  clearFilterOnSelection: PropTypes.bool,
  /**
   * Replace the regular appearance of combo box option with custom renderer.
   */
  optionRenderer: PropTypes.func
};
Combobox.defaultProps = {
  className: "",
  placeholder: "",
  id: "",
  noResultsMessage: "No results found",
  disabled: false,
  options: [],
  categories: undefined,
  withCategoriesDivider: false,
  size: Combobox.sizes.MEDIUM,
  optionLineHeight: 32,
  optionsListHeight: undefined,
  autoFocus: false,
  /**
   * Callback that called after clicking on the add new combo box button.
   */
  onAddNew: undefined,
  /**
   * The button label appears at the end of the combo box when the search does not return appropriate options.
   * The button will not be displayed if you have not passed a function for the onAddNew prop
   */
  addNewLabel: "Add new",
  filter: defaultFilter,
  disableFilter: false,
  onFilterChanged: undefined,
  /** shows loading animation */
  loading: false,
  onOptionHover: NOOP,
  onOptionLeave: NOOP,
  shouldScrollToSelectedItem: true,
  noResultsRenderer: undefined,
  stickyCategories: false,
  optionRenderer: null,
  clearFilterOnSelection: false
};

export default Combobox;
