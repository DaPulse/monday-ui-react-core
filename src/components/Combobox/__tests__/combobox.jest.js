import React from "react";
import renderer from "react-test-renderer";
import Combobox from "../Combobox";

describe("<Combobox />", () => {
  it("renders correctly with empty props", () => {
    const tree = renderer.create(<Combobox />).toJSON();
    expect(tree).toMatchSnapshot();
  });
  it("renders correctly with one option", () => {
    const tree = renderer.create(<Combobox option={[{ id: "1", label: "Test" }]} />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
