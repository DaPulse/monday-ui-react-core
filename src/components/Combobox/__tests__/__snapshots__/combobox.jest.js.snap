// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Combobox /> renders correctly with empty props 1`] = `
<div
  className="combobox--wrapper"
  id=""
>
  <div
    className="input-component search_component_wrapper"
    role="search"
  >
    <div
      className="input-component__label--wrapper"
    >
      <div
        className="input-component__input-wrapper input-component__input-wrapper--size-medium"
      >
        <input
          aria-activedescendant=""
          aria-invalid={null}
          aria-label="Search for content"
          aria-owns=""
          autoComplete="off"
          autoFocus={false}
          className="combobox--wrapper-search search_component input-component__input input-component__input--has-icon"
          disabled={false}
          id="combobox-search"
          maxLength={null}
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          placeholder=""
          readOnly={false}
          required={false}
          role=""
          type="search"
          value=""
        />
        <div
          className="input-component__icon--container input-component__icon--container-has-icon input-component__icon--container-active"
          onClick={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchCancel={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          role="button"
          tabIndex="-1"
          type="button"
        >
          <span
            aria-label="Search"
            className="icon_component input-component__icon icon_component--no-focus-style fa fa-search"
            onClick={[Function]}
            role=""
            tabIndex={-1}
          />
        </div>
        <div
          className="input-component__icon--container input-component__icon--container-has-icon"
          onClick={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchCancel={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          role="button"
          tabIndex="-1"
          type="button"
        >
          <span
            aria-label="Clear Search"
            className="icon_component input-component__icon icon_component--no-focus-style fa fa-close"
            onClick={[Function]}
            role=""
            tabIndex={-1}
          />
        </div>
      </div>
    </div>
  </div>
  <div
    className="combobox--wrapper-list"
  />
</div>
`;

exports[`<Combobox /> renders correctly with one option 1`] = `
<div
  className="combobox--wrapper"
  id=""
>
  <div
    className="input-component search_component_wrapper"
    role="search"
  >
    <div
      className="input-component__label--wrapper"
    >
      <div
        className="input-component__input-wrapper input-component__input-wrapper--size-medium"
      >
        <input
          aria-activedescendant=""
          aria-invalid={null}
          aria-label="Search for content"
          aria-owns=""
          autoComplete="off"
          autoFocus={false}
          className="combobox--wrapper-search search_component input-component__input input-component__input--has-icon"
          disabled={false}
          id="combobox-search"
          maxLength={null}
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          placeholder=""
          readOnly={false}
          required={false}
          role=""
          type="search"
          value=""
        />
        <div
          className="input-component__icon--container input-component__icon--container-has-icon input-component__icon--container-active"
          onClick={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchCancel={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          role="button"
          tabIndex="-1"
          type="button"
        >
          <span
            aria-label="Search"
            className="icon_component input-component__icon icon_component--no-focus-style fa fa-search"
            onClick={[Function]}
            role=""
            tabIndex={-1}
          />
        </div>
        <div
          className="input-component__icon--container input-component__icon--container-has-icon"
          onClick={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchCancel={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          role="button"
          tabIndex="-1"
          type="button"
        >
          <span
            aria-label="Clear Search"
            className="icon_component input-component__icon icon_component--no-focus-style fa fa-close"
            onClick={[Function]}
            role=""
            tabIndex={-1}
          />
        </div>
      </div>
    </div>
  </div>
  <div
    className="combobox--wrapper-list"
  />
</div>
`;
